<a href="https://colab.research.google.com/github/ravindyaparami/Statistical-Learning-e20056/blob/main/E20056_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Q. Gaussian Mixture Clustering as Conditional Updating

Consider a dataset
$$
x_1,x_2,\dots,x_n\in\mathbb R^d.
$$
We wish to cluster these observations into $K$ groups. Instead of assigning each point deterministically to a cluster at the beginning, we introduce a latent random variable
$$
C_i\in{1,\dots,K},
$$
where $C_i=k$ means that the observation $x_i$ belongs to cluster $k$.
Let the prior probability of cluster membership be
$$
P(C_i=k)=\phi_k,
$$
where
$$
\phi_k\ge 0,
\qquad
\sum_{k=1}^K \phi_k=1.
$$

Conditional on $C_i=k$, assume that the observation $X_i$ is generated from a multivariate Gaussian distribution:
$$
X_i\mid C_i=k
\sim
\mathscr N(\mu_k,\Sigma_k),
$$
where
$$
\mu_k\in\mathbb R^d,
\qquad
\Sigma_k\in\mathbb R^{d\times d}
$$
are the mean vector and covariance matrix of cluster $k$.

The model parameters
$$
\phi_k,\mu_k,\Sigma_k,
\qquad k=1,\dots,K,
$$
are assumed to be fixed but unknown.

---

1. Deriving the Marginal Density:
Using the law of total probability, show that the marginal density of $X_i$ is
$$
p(x_i)=\sum_{k=1}^K
\phi_k
\mathscr N(x_i\mid \mu_k,\Sigma_k).
$$
Explain why this density is called a Gaussian mixture density.

---

2. Deriving the Posterior Cluster Probability:
For a fixed observation $x_i$, use Bayes' rule to derive
$$
P(C_i=k\mid X_i=x_i)=\frac{
P(X_i=x_i\mid C_i=k)P(C_i=k)
}{
\sum_{j=1}^K P(X_i=x_i\mid C_i=j)P(C_i=j)
}.
$$
Then substitute the Gaussian model and the cluster prior to obtain
$$
P(C_i=k\mid X_i=x_i)=\frac{
\phi_k\mathscr N(x_i\mid \mu_k,\Sigma_k)
}{
\sum_{j=1}^K
\phi_j\mathscr N(x_i\mid \mu_j,\Sigma_j)
}.
$$
This quantity is called the responsibility of cluster $k$ for data point $x_i$, and is denoted by
$$
\gamma_{ik}=P(C_i=k\mid X_i=x_i).
$$
Explain why $\gamma_{ik}$ may be interpreted as a posterior probability of cluster membership.

---

3. One-Hot Encoding of the Latent Cluster Variable:
Now define a one-hot encoded latent random vector
$$
Z_i=
\begin{bmatrix}
Z_{i1}\\
Z_{i2}\\
\vdots\\
Z_{iK}
\end{bmatrix},
$$
where
$$
Z_{ik}=\begin{cases}
1, & \text{if } C_i=k,\\
0, & \text{otherwise}.
\end{cases}
$$
Show that
$$
\mathbb E[Z_{ik}\mid X_i=x_i]=P(C_i=k\mid X_i=x_i).
$$
Hence show that
$$
\mathbb E[Z_i\mid X_i=x_i]=\begin{bmatrix}
\gamma_{i1}\\
\gamma_{i2}\\
\vdots\\
\gamma_{iK}
\end{bmatrix}.
$$
Conclude that the soft cluster assignment in a Gaussian mixture model is precisely the conditional expectation
$$
\mathbb E[Z_i\mid X_i=x_i].
$$

---

4. From Soft Assignment to Hard Clustering:
The vector
$$
\mathbb E[Z_i\mid X_i=x_i]
$$
gives a soft assignment of $x_i$ to all clusters. A hard cluster assignment can be obtained by choosing the cluster with the largest posterior probability:
$$
\widehat C_i=\operatorname{arg\,max}_{1\le k\le K}
\gamma_{ik}.
$$
Explain the difference between soft clustering and hard clustering in this context.

---

5. Conditional Expectation of the Observation Given the Cluster:
Show that
$$
\mathbb E[X_i\mid C_i=k]=\mu_k.
$$
Explain why $\mu_k$ can be interpreted as the center of cluster $k$.
Then compare the two conditional expectations
$$
\mathbb E[Z_i\mid X_i=x_i]
$$
and
$$
\mathbb E[X_i\mid C_i=k].
$$
Explain why the first gives the soft cluster membership of an observed point, while the second gives the mean location of a cluster.

---

6. The Complete-Data Likelihood
If the latent labels $z_i$ were known, the complete-data likelihood would be
$$
p(x_1,\dots,x_n,z_1,\dots,z_n)=\prod_{i=1}^n
\prod_{k=1}^K
\left[
\phi_k
\mathscr N(x_i\mid \mu_k,\Sigma_k)
\right]^{z_{ik}}.
$$
Take the logarithm and show that the complete-data log-likelihood is
$$
\ell_c=\sum_{i=1}^n
\sum_{k=1}^K
z_{ik}
\left[
\log \phi_k
+
\log \mathscr N(x_i\mid \mu_k,\Sigma_k)
\right].
$$
Explain why this expression would be easy to maximize if the values of $z_{ik}$ were known.

---

7. The EM Interpretation:
In practice, the latent variables $Z_i$ are not observed. The EM algorithm replaces the unknown indicators $z_{ik}$ by their conditional expectations given the observed data and current parameter estimates:
$$
z_{ik}
\quad\leadsto\quad
\mathbb E[Z_{ik}\mid X_i=x_i].
$$
That is,
$$
z_{ik}
\quad\leadsto\quad
\gamma_{ik}.
$$
This is the E-step of the EM algorithm.
Using this idea, write the expected complete-data log-likelihood:
$$
Q=\sum_{i=1}^n
\sum_{k=1}^K
\gamma_{ik}
\left[
\log \phi_k
+
\log \mathscr N(x_i\mid \mu_k,\Sigma_k)
\right].
$$
Explain why the E-step can be interpreted as a conditional update of cluster membership probabilities.

---

8. Parameter Updates:
By maximizing $Q$ with respect to the model parameters, derive the standard GMM updates:
$$
N_k=\sum_{i=1}^n \gamma_{ik},
$$
$$
\phi_k^{\text{new}}=\frac{N_k}{n},
$$
$$
\mu_k^{\text{new}}=\frac{1}{N_k}
\sum_{i=1}^n
\gamma_{ik}x_i,
$$
and
$$
\Sigma_k^{\text{new}}=\frac{1}{N_k}
\sum_{i=1}^n
\gamma_{ik}
(x_i-\mu_k^{\text{new}})
(x_i-\mu_k^{\text{new}})^T.
$$
Explain how the responsibility $\gamma_{ik}$ acts as a fractional membership weight of observation $x_i$ in cluster $k$.

---

9. Interpretation:
Write a short paragraph explaining why GMM clustering can be viewed as a repeated process of conditional updating.
Your answer should mention the following points:

* The mixture weight $\phi_k$ is the prior probability of cluster $k$.
* The Gaussian density $\mathscr N(x_i\mid \mu_k,\Sigma_k)$ measures how compatible $x_i$ is with cluster $k$.
* The responsibility $\gamma_{ik}$ is the posterior probability of cluster $k$ after observing $x_i$.
* The soft assignment vector is
$$
\mathbb E[Z_i\mid X_i=x_i].
$$

* The M-step updates the cluster parameters using these posterior membership probabilities as weights.
Conclude that Gaussian mixture clustering is probabilistic clustering based on conditional expectations of latent cluster membership variables.

---

Here is a perfectly tailored question that you can add as the final part (**Part 10**) of your assignment notebook to bridge your theoretical derivations with your code implementation:

---

10. Computational Simulation and Out-of-Sample Validation

Using the theoretical framework established in the previous parts, write a Python class named `GMMFinancialSegmenter` that implements a two-dimensional Gaussian Mixture Model (GMM) using `scikit-learn` and visualizes the results interactively using `Plotly`. Your implementation should fulfill the following criteria:

* **Data Splitting and Scaling:** Accept a dataset containing two continuous features (e.g., mimicking financial behaviors like `PURCHASES` and `CREDIT_LIMIT`), standardize the features to handle variance scaling, and split the data into an 80% training set and a 20% validation/test set.
* **EM Execution:** Fit a GMM with $K=3$ components on the training data using the Expectation-Maximization (EM) algorithm, printing whether the model successfully converged and the number of iterations required.
* **Out-of-Sample Performance:** Compute and output the average log-likelihood score over the unseen test set to validate how well the learned density functions generalize to new data.
* **Interactive Visualizations:** Implement methods to generate three distinct Plotly figures:
1. An empirical **2D Density Heatmap** of the raw training data with marginal distributions to inspect its underlying multimodal structure.
2. A **Training Assignment Plot** that overlays the training data points on top of a continuous contour map showing the maximum posterior responsibilities ($\gamma_{ik}$) computed across a fine coordinate grid.
3. A **Test Assignment Plot** that replicates the contour boundary visualization but overlays out-of-sample test data points to expose the physical regions of cluster ambiguity.



Briefly evaluate the resulting plots. Explain how the continuous background contour map visually demonstrates the soft assignment expectation vector $\mathbb{E}[Z_i \mid X_i = x_{\text{grid}}]$ that you proved analytically in Part 3.

Use the dataset

https://www.kaggle.com/datasets/arjunbhasin2013/ccdata

#Answers

Gaussian Mixture Clustering as Conditional Updating

Consider observations

$$
x_1,x_2,\ldots,x_n\in\mathbb{R}^d
$$

that we wish to divide into $K$ clusters.

Introduce a latent cluster variable

$$
C_i\in{1,\ldots,K},
$$

where

$$
C_i=k
$$

means that observation $x_i$ belongs to cluster $k$.

The prior probability of cluster membership is

$$
P(C_i=k)=\phi_k,
$$

where

$$
\phi_k\geq0,
\qquad
\sum_{k=1}^{K}\phi_k=1.
$$

Conditional on belonging to cluster $k$, assume

$$
X_i\mid C_i=k
\sim
\mathcal{N}(\mu_k,\Sigma_k),
$$

where

$$
\mu_k\in\mathbb{R}^d
$$

is the mean of cluster $k$ and

$$
\Sigma_k\in\mathbb{R}^{d\times d}
$$

is its covariance matrix.

1. Deriving the Marginal Density

Using the law of total probability, the density of $X_i$ can be obtained by summing over all possible latent clusters:

\sum_{k=1}^{K}
p(x_i,C_i=k).
$$

Using

p(x_i\mid C_i=k)P(C_i=k),
$$

we obtain

\sum_{k=1}^{K}
P(C_i=k)
p(x_i\mid C_i=k).
$$

Since

$$
P(C_i=k)=\phi_k
$$

and

\mathcal{N}(x_i\mid\mu_k,\Sigma_k),
$$

the marginal density becomes

\sum_{k=1}^{K}
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
}
$$

where the multivariate Gaussian density is

\frac{1}
{(2\pi)^{d/2}|\Sigma_k|^{1/2}}
\exp
\left[
-\frac12
(x_i-\mu_k)^T
\Sigma_k^{-1}
(x_i-\mu_k)
\right].
$$

Why is this called a Gaussian mixture density?

The marginal density is a weighted sum of $K$ Gaussian densities.

Each Gaussian represents one component or cluster, while $\phi_k$ specifies the relative contribution of component $k$.

Therefore,

\phi_1\mathcal{N}_1(x)
+
\phi_2\mathcal{N}_2(x)
+\cdots+
\phi_K\mathcal{N}_K(x).
$$

It is therefore called a Gaussian Mixture Model (GMM).

Unlike a single Gaussian distribution, a Gaussian mixture can represent:

multiple modes or peaks,
clusters with different centers,
clusters with different variances,
clusters with different orientations and correlations.
2. Deriving the Posterior Cluster Probability

For a fixed observed point $x_i$, we want to determine the probability that it belongs to cluster $k$:

$$
P(C_i=k\mid X_i=x_i).
$$

Using Bayes' rule,

\frac{
p(x_i\mid C_i=k)P(C_i=k)
}{
p(x_i)
}.
$$

Since

\sum_{j=1}^{K}
p(x_i\mid C_i=j)P(C_i=j),
$$

we obtain

\frac{
p(x_i\mid C_i=k)P(C_i=k)
}{
\displaystyle
\sum_{j=1}^{K}
p(x_i\mid C_i=j)P(C_i=j)
}.
$$

For continuous $X_i$, the terms involving $X_i=x_i$ are interpreted as probability densities, rather than probabilities at a single point.

Substituting

$$
P(C_i=k)=\phi_k
$$

and

\mathcal{N}(x_i\mid\mu_k,\Sigma_k),
$$

gives

\frac{
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
}{
\displaystyle
\sum_{j=1}^{K}
\phi_j
\mathcal{N}(x_i\mid\mu_j,\Sigma_j)
}
}
$$

This probability is called the responsibility of cluster $k$ for observation $x_i$:

P(C_i=k\mid X_i=x_i)
}
$$

The responsibilities satisfy

$$
0\leq\gamma_{ik}\leq1
$$

and

$$
\sum_{k=1}^{K}\gamma_{ik}=1.
$$

Interpretation

The responsibility $\gamma_{ik}$ is a posterior probability because it combines:

Prior cluster probability

$$
\phi_k=P(C_i=k)
$$

with the likelihood

$$
\mathcal{N}(x_i\mid\mu_k,\Sigma_k).
$$

Thus,

$$
\text{Posterior}
\propto
\text{Prior}
\times
\text{Likelihood}.
$$

A large $\gamma_{ik}$ means that, after observing $x_i$, there is strong evidence that the point belongs to cluster $k$.

3. One-Hot Encoding of the Latent Cluster Variable

Define the one-hot latent vector

\begin{bmatrix}
Z_{i1}\
Z_{i2}\
\vdots\
Z_{iK}
\end{bmatrix},
$$

where

\begin{cases}
1, & C_i=k,\
0, & \text{otherwise}.
\end{cases}
$$

Because $Z_{ik}$ is an indicator random variable,

$$
E[Z_{ik}\mid X_i=x_i]
$$

can be written as

1\cdot P(Z_{ik}=1\mid X_i=x_i)
+
0\cdot P(Z_{ik}=0\mid X_i=x_i).
$$

Therefore,

P(Z_{ik}=1\mid X_i=x_i).
$$

But

$$
Z_{ik}=1
\iff
C_i=k.
$$

Hence,

P(C_i=k\mid X_i=x_i)

\gamma_{ik}
}
$$

Therefore,

\begin{bmatrix}
\gamma_{i1}\
\gamma_{i2}\
\vdots\
\gamma_{iK}
\end{bmatrix}
}
$$

Thus, the soft cluster assignment vector is precisely

$$
\boxed{
E[Z_i\mid X_i=x_i]
}
$$

This is an important interpretation of GMM clustering: instead of immediately assigning a point to one cluster, the model computes the expected values of its latent cluster membership indicators.

4. From Soft Assignment to Hard Clustering

The soft assignment for observation $x_i$ is

\begin{bmatrix}
\gamma_{i1}\
\gamma_{i2}\
\vdots\
\gamma_{iK}
\end{bmatrix}.
$$

For example,

\begin{bmatrix}
0.10\
0.65\
0.25
\end{bmatrix}
$$

means that the observation has:

10% posterior probability of belonging to cluster 1,
65% posterior probability of belonging to cluster 2,
25% posterior probability of belonging to cluster 3.

This is called soft clustering because uncertainty about cluster membership is retained.

A hard assignment is obtained using

\arg\max_{1\leq k\leq K}
\gamma_{ik}
}
$$

For the previous example,

$$
\hat C_i=2.
$$

Difference between soft and hard clustering
Soft clustering

Soft clustering preserves all posterior probabilities:

$$
(\gamma_{i1},\ldots,\gamma_{iK}).
$$

It therefore represents uncertainty and overlapping clusters.

Hard clustering

Hard clustering assigns each point to only the cluster with the largest responsibility.

It retains only

$$
\arg\max_k\gamma_{ik}
$$

and discards the probabilities of the other clusters.

Therefore, GMM is fundamentally a soft probabilistic clustering method, although hard labels can be generated from its posterior probabilities.

5. Conditional Expectation of the Observation Given the Cluster

By assumption,

$$
X_i\mid C_i=k
\sim
\mathcal{N}(\mu_k,\Sigma_k).
$$

The expected value of a Gaussian random vector is its mean.

Therefore,

\mu_k
}
$$

Hence, $\mu_k$ represents the expected location of observations generated from cluster $k$.

For this reason, $\mu_k$ can be interpreted as the center of cluster $k$.

Comparison of the Two Conditional Expectations

The first conditional expectation is

\begin{bmatrix}
\gamma_{i1}\
\vdots\
\gamma_{iK}
\end{bmatrix}.
$$

This answers:

Given an observed point $x_i$, how likely is it to belong to each cluster?

Therefore, it gives the soft cluster membership of an observed point.

The second conditional expectation is

\mu_k.
$$

This answers:

Given that an observation belongs to cluster $k$, where do we expect that observation to be located?

Therefore, it gives the mean spatial location or center of cluster $k$.

Thus,

$$
\boxed{
E[Z_i\mid X_i=x_i]
\rightarrow
\text{membership information}
}
$$

while

$$
\boxed{
E[X_i\mid C_i=k]
\rightarrow
\text{cluster location information}.
}
$$

6. Complete-Data Likelihood

Suppose that both the observations and their latent labels were known.

The complete-data likelihood is

\prod_{i=1}^{n}
\prod_{k=1}^{K}
\left[
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]^{z_{ik}}.
$$

Taking the logarithm,

\log
\left[
\prod_{i=1}^{n}
\prod_{k=1}^{K}
\left(
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right)^{z_{ik}}
\right].
$$

Using

\sum_j\log a_j,
$$

we obtain

\sum_{i=1}^{n}
\sum_{k=1}^{K}
z_{ik}
\log
\left[
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right].
$$

Using

$$
\log(ab)=\log a+\log b,
$$

the complete-data log-likelihood becomes

\sum_{i=1}^{n}
\sum_{k=1}^{K}
z_{ik}
\left[
\log\phi_k
+
\log
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]
}
$$

Why would this be easy to maximize if $z_{ik}$ were known?

If the indicators were known, we would know exactly which observations belong to each cluster.

For cluster $k$, only observations having

$$
z_{ik}=1
$$

would contribute to its parameter estimates.

We could then directly calculate:

the proportion of observations in each cluster,
the sample mean of each cluster,
the sample covariance of each cluster.

Thus, parameter estimation would reduce to ordinary Gaussian maximum-likelihood estimation for separately labeled groups.

The difficulty is that the $z_{ik}$ values are hidden.

7. EM Interpretation

Since the latent variables $Z_i$ are not observed, the Expectation-Maximization algorithm works with their conditional expectations.

During the E-step,

$$
z_{ik}
\quad\longrightarrow\quad
E[Z_{ik}\mid X_i=x_i].
$$

Since

\gamma_{ik},
$$

we replace

$$
\boxed{
z_{ik}\longrightarrow\gamma_{ik}.
}
$$

The expected complete-data log-likelihood becomes

\sum_{i=1}^{n}
\sum_{k=1}^{K}
\gamma_{ik}
\left[
\log\phi_k
+
\log
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]
}
$$

More precisely, at each EM iteration, the responsibilities are calculated using the current parameter estimates.

The E-step therefore calculates

\frac{
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
}{
\displaystyle
\sum_{j=1}^{K}
\phi_j
\mathcal{N}(x_i\mid\mu_j,\Sigma_j)
}.
$$

Why is the E-step a conditional update?

Before observing $x_i$, the cluster probability is

$$
P(C_i=k)=\phi_k.
$$

After observing $x_i$, this prior belief is updated using Bayes' rule:

$$
\phi_k
\quad\longrightarrow\quad
\gamma_{ik}.
$$

Thus,

$$
\text{Prior membership probability}
+
\text{Observed data}
\rightarrow
\text{Posterior membership probability}.
$$

Therefore, the E-step is a conditional probabilistic update of cluster membership.

8. Derivation of the GMM Parameter Updates

Define the effective number of observations belonging to cluster $k$ as

\sum_{i=1}^{n}
\gamma_{ik}.
}
$$

Because $\gamma_{ik}$ can lie between 0 and 1, $N_k$ is an effective or fractional cluster size.

8.1 Updating the Mixture Weights

The part of $Q$ involving $\phi_k$ is

\sum_{k=1}^{K}
N_k\log\phi_k.
$$

We must maximize this subject to

$$
\sum_{k=1}^{K}\phi_k=1.
$$

Introduce a Lagrange multiplier $\lambda$:

\sum_{k=1}^{K}
N_k\log\phi_k
+
\lambda
\left(
\sum_{k=1}^{K}\phi_k-1
\right).
$$

Differentiate with respect to $\phi_k$:




$$

Therefore,

-\frac{N_k}{\lambda}.
$$

Using

$$
\sum_{k=1}^{K}\phi_k=1
$$

and

$$
\sum_{k=1}^{K}N_k=n,
$$

we obtain

\frac{N_k}{n}.
}
$$

Thus, the updated mixture weight is the effective fraction of observations belonging to cluster $k$.

8.2 Updating the Cluster Mean

The Gaussian contribution involving $\mu_k$ is maximized by solving




$$

Therefore,

\mu_k
\sum_{i=1}^{n}
\gamma_{ik}.
$$

Since

\sum_{i=1}^{n}\gamma_{ik},
$$

we obtain

\frac{1}{N_k}
\sum_{i=1}^{n}
\gamma_{ik}x_i.
}
$$

Therefore, the updated cluster center is a responsibility-weighted mean of all observations.

8.3 Updating the Covariance Matrix

Similarly, maximizing $Q$ with respect to $\Sigma_k$ gives

\frac{1}{N_k}
\sum_{i=1}^{n}
\gamma_{ik}
(x_i-\mu_k^{\text{new}})
(x_i-\mu_k^{\text{new}})^T
}
$$

Therefore, the covariance matrix is a responsibility-weighted covariance around the newly updated mean.

Final M-Step Updates

The complete standard GMM updates are therefore

\sum_{i=1}^{n}\gamma_{ik}
}
$$

\frac{N_k}{n}
}
$$

\frac{1}{N_k}
\sum_{i=1}^{n}
\gamma_{ik}x_i
}
$$

and

\frac{1}{N_k}
\sum_{i=1}^{n}
\gamma_{ik}
(x_i-\mu_k^{\text{new}})
(x_i-\mu_k^{\text{new}})^T.
}
$$

Interpretation of $\gamma_{ik}$ as a Fractional Membership Weight

Unlike hard clustering, GMM does not initially treat an observation as belonging completely to only one cluster.

For example, suppose

(0.20,0.70,0.10).
$$

Then observation $x_i$ contributes:

weight $0.20$ to cluster 1,
weight $0.70$ to cluster 2,
weight $0.10$ to cluster 3.

Therefore, $\gamma_{ik}$ acts as a fractional membership weight when computing the new means, covariances, and mixture proportions.

9. Interpretation: GMM as Repeated Conditional Updating

Gaussian mixture clustering can be viewed as a repeated process of conditional probabilistic updating.

The mixture weight

$$
\phi_k
$$

represents the prior probability that an observation belongs to cluster $k$ before considering its observed feature values.

The Gaussian density

$$
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
$$

measures how compatible the observed point $x_i$ is with cluster $k$.

Bayes' rule combines these quantities to calculate

P(C_i=k\mid X_i=x_i),
$$

which is the posterior probability of cluster membership after observing $x_i$.

The complete soft assignment is

\begin{bmatrix}
\gamma_{i1}\
\vdots\
\gamma_{iK}
\end{bmatrix}.
$$

The E-step calculates these conditional expectations using the current cluster parameters.

The M-step then uses the responsibilities as fractional membership weights to update

$$
\phi_k,\qquad
\mu_k,\qquad
\Sigma_k.
$$

The new parameters change the Gaussian component distributions, which in turn change the responsibilities during the next E-step.

Therefore, EM repeatedly performs

$$
\boxed{
\text{Estimate memberships}
\rightarrow
\text{Update cluster parameters}
\rightarrow
\text{Re-estimate memberships}
\rightarrow\cdots
}
$$

until the model converges.

Hence, Gaussian mixture clustering is a form of probabilistic clustering based on conditional expectations of latent cluster membership variables.

10. Computational Simulation and Out-of-Sample Validation

We now implement a two-dimensional GMM using the credit-card customer dataset.

The two selected continuous variables are:

PURCHASES: customer purchase amount.
CREDIT_LIMIT: customer's credit limit.

The purpose is to fit a

$$
K=3
$$

component Gaussian mixture model and connect the computational results with the theoretical responsibilities

P(C_i=k\mid X_i=x_i).
$$

In [ ]:
!pip -q install kagglehub

In [ ]:
import kagglehub
import pandas as pd
from pathlib import Path

# Download Kaggle dataset
dataset_path = kagglehub.dataset_download( "arjunbhasin2013/ccdata" )
print("Dataset downloaded to:")
print(dataset_path)

# Find the CSV file
csv_files = list(Path(dataset_path).glob("*.csv"))
if not csv_files:
  raise FileNotFoundError(
      "No CSV file was found in the downloaded dataset." )

# Prefer CC GENERAL.csv if available
csv_path = next( (
    file for file in csv_files
    if file.name.lower() == "cc general.csv"
    ),
    csv_files[0] )

print("\nUsing file:")
print(csv_path)

# Load data
df = pd.read_csv(csv_path)
print("\nDataset shape:", df.shape)
display(df.head())
print("\nSelected feature summary:")
display( df[["PURCHASES", "CREDIT_LIMIT"]].describe() )
print("\nMissing values:")
print( df[["PURCHASES", "CREDIT_LIMIT"]] .isnull() .sum() )

Using Colab cache for faster access to the 'ccdata' dataset.
Dataset downloaded to:
/kaggle/input/ccdata

Using file:
/kaggle/input/ccdata/CC GENERAL.csv

Dataset shape: (8950, 18)


,CUST_ID,BALANCE,BALANCE_FREQUENCY,PURCHASES,ONEOFF_PURCHASES,INSTALLMENTS_PURCHASES,CASH_ADVANCE,PURCHASES_FREQUENCY,ONEOFF_PURCHASES_FREQUENCY,PURCHASES_INSTALLMENTS_FREQUENCY,CASH_ADVANCE_FREQUENCY,CASH_ADVANCE_TRX,PURCHASES_TRX,CREDIT_LIMIT,PAYMENTS,MINIMUM_PAYMENTS,PRC_FULL_PAYMENT,TENURE
0,C10001,40.900749,0.818182,95.40,0.00,95.4,0.000000,0.166667,0.000000,0.083333,0.000000,0,2,1000.0,201.802084,139.509787,0.000000,12
1,C10002,3202.467416,0.909091,0.00,0.00,0.0,6442.945483,0.000000,0.000000,0.000000,0.250000,4,0,7000.0,4103.032597,1072.340217,0.222222,12
2,C10003,2495.148862,1.000000,773.17,773.17,0.0,0.000000,1.000000,1.000000,0.000000,0.000000,0,12,7500.0,622.066742,627.284787,0.000000,12
3,C10004,1666.670542,0.636364,1499.00,1499.00,0.0,205.788017,0.083333,0.083333,0.000000,0.083333,1,1,7500.0,0.000000,NaN,0.000000,12
4,C10005,817.714335,1.000000,16.00,16.00,0.0,0.000000,0.083333,0.083333,0.000000,0.000000,0,1,1200.0,678.334763,244.791237,0.000000,12



Selected feature summary:


,PURCHASES,CREDIT_LIMIT
count,8950.000000,8949.000000
mean,1003.204834,4494.449450
std,2136.634782,3638.815725
min,0.000000,50.000000
25%,39.635000,1600.000000
50%,361.280000,3000.000000
75%,1110.130000,6500.000000
max,49039.570000,30000.000000



Missing values:
PURCHASES       0
CREDIT_LIMIT    1
dtype: int64


10.2 GMMFinancialSegmenter Class

The class below performs:

Missing-value removal.

80% training and 20% test splitting.

Standardization using training-set statistics only.

GMM fitting with $K=3$.

EM convergence reporting.

Out-of-sample average log-likelihood evaluation.

Responsibility calculation.

Three interactive Plotly visualizations.

In [ ]:
  import numpy as np
  import pandas as pd
  import plotly.express as px
  import plotly.graph_objects as go
  from sklearn.model_selection import train_test_split
  from sklearn.preprocessing import StandardScaler
  from sklearn.mixture import GaussianMixture

  class GMMFinancialSegmenter:
    def __init__( self, n_components=3, test_size=0.20, random_state=42 ):
      """
      Two-dimensional Gaussian Mixture Model segmenter.
      """
      self.n_components = n_components
      self.test_size = test_size
      self.random_state = random_state
      self.features = None
      self.scaler = StandardScaler()
      self.gmm = GaussianMixture(
          n_components=n_components,
          covariance_type="full",
          n_init=10,
          max_iter=500,
          reg_covar=1e-6,
          random_state=random_state )

      self.X_train_raw = None
      self.X_test_raw = None
      self.X_train_scaled = None
      self.X_test_scaled = None
      self.train_labels = None
      self.test_labels = None
      self.train_responsibilities = None
      self.test_responsibilities = None
      self.test_log_likelihood = None

    # --------------------------------------------------
    # 1. Prepare data
    # --------------------------------------------------

    def prepare_data( self, dataframe, features=(
        "PURCHASES", "CREDIT_LIMIT") ):
      self.features = list(features)

      # Check columns
      missing_columns = [
          col for col in self.features if col not in dataframe.columns
          ]

      if missing_columns:
        raise ValueError(
            f"Missing required columns: " f"{missing_columns}" )

      # Select only the two required features
      data = (
          dataframe[self.features]
          .replace([np.inf, -np.inf], np.nan)
          .dropna() .astype(float) )

      print( "Number of usable observations:", len(data) )

          # IMPORTANT:
          # Split before fitting the scaler to avoid
          # information leakage from the test data.
      (
            self.X_train_raw,
            self.X_test_raw

        ) = train_test_split(
            data,
            test_size=self.test_size,
            random_state=self.random_state,
            shuffle=True )

      # Fit scaler ONLY on training data
      self.X_train_scaled = (
          self.scaler.fit_transform(
              self.X_train_raw ) )

      # Apply the same transformation to test data
      self.X_test_scaled = (
          self.scaler.transform( self.X_test_raw ) )

      print( "Training observations:", len(self.X_train_raw) )
      print( "Test observations:", len(self.X_test_raw) )
      return self

    # --------------------------------------------------
    # 2. Fit GMM using EM
    # --------------------------------------------------
    def fit(self):
      if self.X_train_scaled is None:
        raise RuntimeError( "Call prepare_data() before fit()." )

      self.gmm.fit( self.X_train_scaled )

      # Hard assignments
      self.train_labels = self.gmm.predict( self.X_train_scaled )
      self.test_labels = self.gmm.predict( self.X_test_scaled )

      # Soft assignments / responsibilities
      self.train_responsibilities = ( self.gmm.predict_proba(
          self.X_train_scaled ) )
      self.test_responsibilities = ( self.gmm.predict_proba(
          self.X_test_scaled ) )

      # Average log-likelihood on unseen data
      self.test_log_likelihood = ( self.gmm.score( self.X_test_scaled ) )

      print( "EM converged:", self.gmm.converged_ )
      print( "Number of EM iterations:", self.gmm.n_iter_ )
      print( "Average test log-likelihood:", round(
          self.test_log_likelihood, 4 ) )
      print("\nLearned mixture weights:")

      for k, weight in enumerate( self.gmm.weights_, start=1 ):
        print( f"Cluster {k}: " f"{weight:.4f}" )
      return self

    # --------------------------------------------------
    # 3. Empirical density heatmap
    # --------------------------------------------------
    def plot_empirical_density(self):
      if self.X_train_raw is None:
        raise RuntimeError( "Data has not been prepared." )

      fig = px.density_heatmap(
          self.X_train_raw,
          x=self.features[0],
          y=self.features[1],
          nbinsx=60,
          nbinsy=60,
          marginal_x="histogram",
          marginal_y="histogram",
          title=( "Empirical 2D Density of " "Training Data" ) )

      fig.update_layout( template="plotly_white" )
      return fig

    # --------------------------------------------------
    # 4. Create responsibility grid
    # --------------------------------------------------
    def _responsibility_grid(
        self,
        grid_size=220,
        quantile_limits=(0.01, 0.99) ):

      if self.train_responsibilities is None:
        raise RuntimeError( "Fit the model before " "creating responsibility maps." )
      # Corrected indentation starts here
      combined = pd.concat( [ self.X_train_raw, self.X_test_raw ], axis=0 )
      x_feature = self.features[0]
      y_feature = self.features[1]
      q_low, q_high = quantile_limits
      x_min = combined[ x_feature ].quantile(q_low)
      x_max = combined[ x_feature ].quantile(q_high)
      y_min = combined[ y_feature ].quantile(q_low)
      y_max = combined[ y_feature ].quantile(q_high)

      # Small padding
      x_pad = 0.05 * (x_max - x_min)
      y_pad = 0.05 * (y_max - y_min)
      x_values = np.linspace(
          max(0, x_min - x_pad),
          x_max + x_pad, grid_size )
      y_values = np.linspace(
          max(0, y_min - y_pad),
          y_max + y_pad, grid_size )
      xx, yy = np.meshgrid( x_values, y_values )

      # Grid in original feature units
      grid_raw = pd.DataFrame( {
          x_feature: xx.ravel(),
          y_feature: yy.ravel() } )

      # Transform using SAME training scaler
      grid_scaled = self.scaler.transform( grid_raw )

      # Compute:
      #
      # [γ_i1, γ_i2, ..., γ_iK]
      #
      responsibilities = ( self.gmm.predict_proba( grid_scaled ) )
      # Largest posterior probability
      max_responsibility = ( responsibilities.max(axis=1) )
      # Hard cluster from argmax
      hard_cluster = ( responsibilities.argmax(axis=1) )
      return ( xx, yy, max_responsibility.reshape( xx.shape ), hard_cluster.reshape( xx.shape ) )

    # --------------------------------------------------
    # 5. Common assignment plot
    # --------------------------------------------------

    def _assignment_plot( self, raw_data, labels, title ):
      ( xx, yy, max_gamma, hard_grid ) = self._responsibility_grid()
      fig = go.Figure()

      # Continuous maximum responsibility map
      fig.add_trace(
          go.Contour(
              x=xx[0],
              y=yy[:, 0],
              z=max_gamma,
              contours=dict(
                  start=1 / self.n_components,
                  end=1.0, size=0.05 ),
              colorbar=dict( title="max γ" ),
              opacity=0.75, name=( "Maximum Posterior " "Responsibility" )
          ) )

      # Overlay observed points
      fig.add_trace( go.Scatter(
          x=raw_data[ self.features[0] ],
          y=raw_data[ self.features[1] ],
          mode="markers",
          marker=dict(
              size=5,
              color=labels,
              colorscale="Turbo",
              line=dict( width=0.3 ) ),
          text=[ f"Cluster {c + 1}" for c in labels ],
          hovertemplate=(
              self.features[0]
              + ": %{x:.2f}<br>"
              + self.features[1]
              + ": %{y:.2f}<br>"
              + "%^{text}" + "<extra></extra>" ),
          name="Observed Points" ) )

      fig.update_layout(
          title=title,
          xaxis_title=self.features[0],
          yaxis_title=self.features[1],
          template="plotly_white" )
      return fig


    # --------------------------------------------------
    # 6. Training assignment plot
    # --------------------------------------------------

    def plot_training_assignments(self):
      return self._assignment_plot(
          self.X_train_raw,
          self.train_labels, (
              "Training Data: GMM Assignments "
              "and Maximum Posterior " "Responsibility" ) )

    # --------------------------------------------------
    # 7. Test assignment plot
    # --------------------------------------------------
    def plot_test_assignments(self):
      return self._assignment_plot( self.X_test_raw, self.test_labels, (
          "Out-of-Sample Test Data: "
          "GMM Assignments and "
          "Maximum Posterior Responsibility" ) )

    # --------------------------------------------------
    # 8. Display component means in original units
    # --------------------------------------------------
    def display_cluster_centers(self):
      centers_original = (
          self.scaler.inverse_transform( self.gmm.means_ ) )
      centers = pd.DataFrame( centers_original, columns=self.features )
      centers.index = [
          f"Cluster {k + 1}" for k in range( self.n_components ) ]

      return centers

In [ ]:
  segmenter = GMMFinancialSegmenter(
      n_components=3,
      test_size=0.20,
      random_state=42 )

  segmenter.prepare_data(
      df,
      features=( "PURCHASES", "CREDIT_LIMIT" ) )
  segmenter.fit()

Number of usable observations: 8949
Training observations: 7159
Test observations: 1790
EM converged: True
Number of EM iterations: 20
Average test log-likelihood: -1.6888

Learned mixture weights:
Cluster 1: 0.1041
Cluster 2: 0.4428
Cluster 3: 0.4531


In [ ]:
  cluster_centers = ( segmenter.display_cluster_centers() )
  display(cluster_centers)

,PURCHASES,CREDIT_LIMIT
Cluster 1,4631.641306,9578.341851
Cluster 2,183.625369,2070.207836
Cluster 3,918.775706,5744.381470


In [ ]:
  fig1 = segmenter.plot_empirical_density()
  fig1.show()

  fig2 = (
      segmenter
      .plot_training_assignments()
      )
  fig2.show()

  fig3 = ( segmenter .plot_test_assignments() )
  fig3.show()

TypeError: cannot unpack non-iterable NoneType object

The model calculates

test_score = (
    segmenter.test_log_likelihood
)

print(
    "Average log-likelihood "
    "on unseen test data:",
    test_score
)

Mathematically, the average test log-likelihood is

$$
\frac{1}{n_{\text{test}}}
\sum_{i=1}^{n_{\text{test}}}
\log
\left[
\sum_{k=1}^{K}
\phi_k
\mathcal{N}
(
x_i^{test}
\mid
\mu_k,\Sigma_k
)
\right].
$$

This measures how much probability density the fitted GMM assigns to unseen observations.

A larger, or less negative, average log-likelihood generally means that the fitted density explains the unseen observations better when comparing models on the same data representation.

It should not be interpreted as classification accuracy because clustering has no true class labels in this application.

10.9 Direct Examination of Responsibilities

The first few responsibility vectors can be displayed using:

In [ ]:
  responsibility_df = pd.DataFrame(
      segmenter.train_responsibilities,
      columns=[ "γ_cluster_1", "γ_cluster_2", "γ_cluster_3" ] )
  display( responsibility_df.head(10) )

,γ_cluster_1,γ_cluster_2,γ_cluster_3
0,0.013622,2.767924e-08,0.986378
1,0.048552,1.076604e-11,0.951448
2,0.020617,2.079435e-03,0.977304
3,0.009742,3.113775e-01,0.678881
4,0.028985,4.570849e-10,0.971015
5,0.471778,7.016247e-37,0.528222
6,0.000794,9.722350e-01,0.026971
7,0.011440,9.452975e-02,0.894030
8,0.013368,3.715767e-02,0.949474
9,0.012271,6.283800e-02,0.924891


For every observation,




$$

This can be verified computationally:

print(
    responsibility_df
    .sum(axis=1)
    .head(10)
)

For example, a hypothetical result such as

$$
[0.02,;0.91,;0.07]
$$

indicates strong membership in cluster 2.

In contrast,

$$
[0.42,;0.46,;0.12]
$$

shows uncertainty between clusters 1 and 2.

The hard assignment would choose cluster 2 in both cases, but only the soft responsibility vector preserves the uncertainty in the second case.

10.10 Relationship Between the Contour Map and the Conditional Expectation

From the theoretical derivation,

\begin{bmatrix}
\gamma_{i1}\
\gamma_{i2}\
\gamma_{i3}
\end{bmatrix}.
$$

The computational visualization extends this idea from only the observed points to every coordinate on a fine grid.

For each grid location

$$
x_{\text{grid}},
$$

the fitted GMM calculates

\begin{bmatrix}
\gamma_1(x_{\text{grid}})\
\gamma_2(x_{\text{grid}})\
\gamma_3(x_{\text{grid}})
\end{bmatrix}.
$$

The contour background displays

$$
\max_k
\gamma_k(x_{\text{grid}}).
$$

Therefore, the background is a continuous visualization of the posterior cluster-membership probabilities derived analytically.

High maximum responsibility means that one component strongly dominates:

$$
E[Z\mid X=x]
\approx
[1,0,0]^T,
$$

or a permutation of this vector.

Lower maximum responsibility indicates that probability is distributed across multiple clusters, for example

$$
E[Z\mid X=x]
\approx
[0.45,0.48,0.07]^T.
$$

These lower-confidence areas correspond to soft cluster boundaries or ambiguity regions.

The scatter-point colors show the hard decision

$$
\arg\max_k\gamma_k(x),
$$

whereas the continuous background preserves information about how certain that decision is.

10.11 Overall Evaluation

The three visualizations illustrate three different aspects of Gaussian mixture clustering.

Empirical Density Heatmap

The density heatmap displays the raw distribution of PURCHASES and CREDIT_LIMIT.

It allows us to visually inspect whether the customer population contains multiple regions of concentration and whether a single Gaussian distribution would be too restrictive.

Training Assignment Plot

The training plot shows how the fitted Gaussian components divide the feature space.

Individual observations receive hard labels for visualization, but the background demonstrates that the underlying GMM calculation is probabilistic.

The transition regions between strongly assigned areas demonstrate that GMM does not fundamentally create rigid boundaries. Instead, posterior membership probabilities vary continuously across the feature space.

Test Assignment Plot

The test plot applies the learned Gaussian mixture to observations that were not used during model training.

This demonstrates out-of-sample conditional updating:

$$
x_{\text{new}}
\rightarrow
P(C=k\mid X=x_{\text{new}}).
$$

A new customer's financial measurements can therefore be inserted into the trained model, and the model calculates a responsibility vector without refitting the GMM.

The average test log-likelihood provides an additional quantitative measure of how well the learned probability density generalizes to the unseen data.

Final Conclusion

Gaussian Mixture Model clustering can be understood naturally as repeated conditional updating of latent cluster memberships.

Before observing a data point, the mixture weight

$$
\phi_k
$$

acts as the prior probability of membership in cluster $k$.

The Gaussian density

$$
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
$$

measures the compatibility of the observed point with that cluster.

Bayes' rule combines these quantities to obtain

P(C_i=k\mid X_i=x_i),
$$

the posterior cluster probability.

Because

\gamma_{ik},
$$

the complete soft assignment is

[
\gamma_{i1},
\ldots,
\gamma_{iK}
]^T.
}
$$

The E-step calculates these posterior membership probabilities.

The M-step then updates

$$
\phi_k,\qquad
\mu_k,\qquad
\Sigma_k
$$

using the responsibilities as fractional weights.

These two steps are repeated until convergence.

The computational contour map makes this theoretical result visible. Every point in the two-dimensional feature space has a complete posterior membership vector, while the hard cluster label simply selects its largest component.

Therefore,

$$
\boxed{
\text{GMM clustering is probabilistic clustering based on conditional expectations of latent cluster membership variables.}
}
$$